docs/intro-to-fuzzing.md

```diff
@@ -0,0 +1,143 @@
+various memory safety issues such as use-after-free bugs and buffer overflows.
+If nothing else, we strongly recommend building any of your fuzzing binaries
+with this instrumentation.
+* [UndefinedBehaviorSanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html)
+detects various forms of
+[undefined behavior](https://en.wikipedia.org/wiki/Undefined_behavior). For
+example, it can detect signed integer overflow, use of misaligned pointers, and
+much more.
+* [MemorySanitizer](https://clang.llvm.org/docs/MemorySanitizer.html)
+detects reads of uninitialized memory. While it is a very useful tool, it can be
+much more difficult to use than the others because it requires that all
+dependencies are also instrumented. We recommend setting this up after you
+already have some familiarity with the other tools.
+
+By taking advantage of these, many subtle bugs will be detected as deterministic
+crashes. This not only allows you to find the bugs more easily, but also makes
+it simpler to reproduce and fix potential issues. These can be used with any of
+the fuzzing tools mentioned in this document, and we strongly recommend taking
+advantage of them.
+
+## What to fuzz?
+
+The [Why Fuzz?](https://github.com/google/fuzzing/blob/master/docs/why-fuzz.md)
+document goes into detail about the types of bugs that fuzzing can catch, and
+the types of applications that may benefit from fuzzing. After deciding to fuzz
+an application, it’s important to identify how that application can be fuzzed.
+
+Any application code exposed to untrusted user input is usually a good candidate
+for fuzzing. Identifying parts of your code that are exposed to untrusted input
+can sometimes be difficult for large applications, but we recommend finding ways
+to define entry points that can be used for your fuzzers, mimicking the types of
+input your application may be exposed to.
+
+For example, if a browser defines a function that takes an input string of CSS
+and parses it, that may be a good candidate for fuzzing. It’s a clear entry
+point that could be used for a fuzzer, and it’s likely that arbitrary
+attacker-supplied input would be handled by it as well.
+
+If code is exposed to the network, it is usually safe to assume it expects to
+handle untrusted inputs. Though the same advice from above applies, some fuzzing
+tools allow this code to be
+[tested directly](https://github.com/google/honggfuzz/blob/master/socketfuzzer/README.md).
+
+## Fuzzing Tools
+
+### [libFuzzer]
+
+[libFuzzer] is a tool which allows you to write fuzz tests in a similar style to
+unit tests. A libFuzzer-based fuzzer could be as simple as the following:
+
+```cpp
+// fuzz_target.cc
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
+  DoSomethingInterestingWithMyAPI(Data, Size);
+  return 0;  // Non-zero return values are reserved for future use.
+}
+```
+
+For new projects setting up fuzzing for the first time, we tend to recommend
```

Comment:
It seems a bit confusing which one someone new to fuzzing should pick. 

Probably make this point stronger about how libFuzzer is the most friendly to use for developers, and should be picked over others. AFL/honggfuzz might be more useful at first to security folks. 

Maybe also talk about how each engine has its own benefits and link to fuzzbench under further reading or somewhere similar?